### PR TITLE
[SPARK-49467][SS] Add support for state data source reader and list state

### PR DESCRIPTION
**Pull request by anishshri-db**
### What changes were proposed in this pull request?
Add support for state data source reader and list state

### Why are the changes needed?
This change adds support for reading state written using list state used primarily within the stateful processor used with the `transformWithState` operator

### Does this PR introduce _any_ user-facing change?
Yes

Users can read state and `explode` entries using the following query:
```
        val stateReaderDf = spark.read
          .format("statestore")
          .option(StateSourceOptions.PATH, <checkpoint_location>)
          .option(StateSourceOptions.STATE_VAR_NAME, <state_var_name>)
          .load()

        val listStateDf = stateReaderDf
          .selectExpr(
            "key.value AS groupingKey",
            "list_value AS valueList",
            "partition_id")
          .select($"groupingKey",
            explode($"valueList").as("valueList"))
```

### How was this patch tested?
Added unit tests

```
[info] Run completed in 1 minute, 3 seconds.
[info] Total number of tests run: 8
[info] Suites: completed 1, aborted 0
[info] Tests: succeeded 8, failed 0, canceled 0, ignored 0, pending 0
[info] All tests passed.
```

### Was this patch authored or co-authored using generative AI tooling?
No

### CODE REVIEW

**Comment by anishshri-db** [comment]
cc - @HeartSaVioR @jingz-db - could you PTAL ? Thx

**Review by jingz-db — APPROVED** [review]
Looks good! Thanks for making the change. Left some small nits.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/StatePartitionReader.scala [hunk 1]
```diff
@@ -68,10 +69,18 @@ abstract class StatePartitionReaderBase(
     stateVariableInfoOpt: Option[TransformWithStateVariableInfo],
     stateStoreColFamilySchemaOpt: Option[StateStoreColFamilySchema])
   extends PartitionReader[InternalRow] with Logging {
+  protected val schemaForValueRow: StructType =
```
Comment (jingz-db) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/StatePartitionReader.scala:72]: Nits: we can use `private` as this is not used in the inherited class.

Comment (jingz-db) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/StatePartitionReader.scala:72]: Also perhaps we can leave a small comment here that we are only using this for state variables dummy value state in `createAndInit()`.

Comment (anishshri-db) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/StatePartitionReader.scala]: Done

**Comment by anishshri-db** [comment]
@HeartSaVioR - could you PTAL ? thx

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/StatePartitionReader.scala [hunk 2]
```diff
@@ -166,16 +184,22 @@ class StatePartitionReader(
         stateVariableInfoOpt match {
           case Some(stateVarInfo) =>
             val stateVarType = stateVarInfo.stateVariableType
-            val hasTTLEnabled = stateVarInfo.ttlEnabled
```
Comment (HeartSaVioR) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/StatePartitionReader.scala:169]: Is it intentional to remove out the readability of TTL functionality for value state? Just wanted to know whether we give up functionality in certain reason, or you want to defer this till we have every types in support range and think about UX which works for every types.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/utils/SchemaUtil.scala [hunk 3]
```diff
@@ -91,23 +90,22 @@ object SchemaUtil {
...
 
     val expectedFieldNames = if (sourceOptions.readChangeFeed) {
       Seq("batch_id", "change_type", "key", "value", "partition_id")
     } else if (transformWithStateVariableInfoOpt.isDefined) {
       val stateVarInfo = transformWithStateVariableInfoOpt.get
-      val hasTTLEnabled = stateVarInfo.ttlEnabled
       val stateVarType = stateVarInfo.stateVariableType
 
       stateVarType match {
         case StateVariableType.ValueState =>
```
Comment (HeartSaVioR) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/utils/SchemaUtil.scala:104]: Now the schema is quite dynamic - have we run `explain` to check the schema is captured in dry-run as well?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/StatePartitionReader.scala [hunk 4]
```diff
@@ -166,16 +184,22 @@ class StatePartitionReader(
...
-                  SchemaUtil.unifyStateRowPairWithTTL((pair.key, pair.value), valueSchema,
-                    partition.partition)
-                } else {
-                  SchemaUtil.unifyStateRowPair((pair.key, pair.value), partition.partition)
+                SchemaUtil.unifyStateRowPair((pair.key, pair.value), partition.partition)
+
+              case StateVariableType.ListState =>
+                val key = pair.key
+                val result = store.valuesIterator(key, stateVarName)
+                var unsafeRowArr: Seq[UnsafeRow] = Seq.empty
```
Comment (HeartSaVioR) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/StatePartitionReader.scala:195]: Worth noting that we are materializing every element in the list, so despite we avoid memory issue in ListState, the problem will pop up when reading ListState via state data source reader.

I guess it's probably worth reconsidering the schema and UX. e.g. having `index` and `value` as columns for the value of list state and allow multiple rows for the same state value. (You've used explode in the example but we could just provide that result directly.) You'll also have the same issue with MapType as well, so worth considering.

**Review by HeartSaVioR — COMMENTED** [review]
First pass, didn't go through test as we'd probably need to reconsider the UX and it's better to review the test later.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/StatePartitionReader.scala (hunk 2 above)
Comment (anishshri-db) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/StatePartitionReader.scala:169]: Yea - for list state, we can't have a separate column. So in that case, we have to embed it as part of the value itself. So trying to make it uniform for value state as well

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/utils/SchemaUtil.scala (hunk 3 above)
Comment (anishshri-db) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/utils/SchemaUtil.scala:104]: Not sure what you mean. The only difference is the additional sub-column if TTL is enabled right ?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/StatePartitionReader.scala (hunk 4 above)
Comment (anishshri-db) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/StatePartitionReader.scala:195]: For this, I was thinking of adding like a `maxEntriesPerRow` type setting (sort of like a pagination thing) which would bound the list/map entries for the same grouping key, but the schema would remain `ArrayType/MapType` per row. Was thinking of adding this as a separate change though after the base change was merged

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/utils/SchemaUtil.scala (hunk 3 above)
Comment (HeartSaVioR) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/utils/SchemaUtil.scala:104]: No, what I meant is whether inferSchema could capture all the details. I assume so but wanted to double confirm with the result of running "explain" for value state vs list state.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/StatePartitionReader.scala (hunk 2 above)
Comment (HeartSaVioR) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/StatePartitionReader.scala:169]: But say, if we explode elements in list state (and map state), we will be able to add TTL column for them and the schema for the value would be very similar (only user key in map state). Do I understand correctly?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/StatePartitionReader.scala (hunk 4 above)
Comment (HeartSaVioR) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/StatePartitionReader.scala:195]: I think it's even OK to not have index and just have multiple rows for the same state key. (It's same as explode) Having index would help us to look into detail of how array elements are stored though.

But I don't feel like having 1000s of elements (even not primitive) in ArrayType is great in overall.

Comment (HeartSaVioR) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/StatePartitionReader.scala:195]: cc. @cloud-fan I'd like to get some advice from Spark SQL expert. Could you please take a quick look into comments and provide some input? Thanks!

Comment (HeartSaVioR) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/StatePartitionReader.scala:195]: Discussed with @cloud-fan . We came up with conclusion that there is no good solution. 

I'm OK with this - we have more time to revisit this (addition of TTL, etc.) and change our mind later if we think there is a better way.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/StatePartitionReader.scala (hunk 2 above)
Comment (HeartSaVioR) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/StatePartitionReader.scala:169]: As commented in other comment thread, I'm OK with this change, but let's see how we could deal with TTL in general. We can defer the decision till we are seeing a full picture of this (among state types).

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/utils/SchemaUtil.scala (hunk 3 above)
Comment (HeartSaVioR) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/utils/SchemaUtil.scala:104]: Please consider this as post-review comment. Not a blocker for merging this PR.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/StatePartitionReader.scala (hunk 2 above)
Comment (HeartSaVioR) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/StatePartitionReader.scala:169]: OK, I see what you are doing now. It's just showing the value schema as passthrough, so TTL will be just a part of value if it's described. OK.

**Review by HeartSaVioR — APPROVED** [review]
+1

**Comment by HeartSaVioR** [comment]
Thanks! Merging to master.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/StatePartitionReader.scala (hunk 2 above)
Comment (anishshri-db) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/StatePartitionReader.scala:169]: Correct yea

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/StatePartitionReader.scala (hunk 4 above)
Comment (anishshri-db) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/StatePartitionReader.scala:195]: Yea sounds good - we'll discuss offline and update the UX as needed

**Pull request closed without merging**